File: apps/ios/SignalASI/AgentPlanEditor.swift
import Foundation

struct AgentPlanEditResult: Equatable {
  var plan: AgentPlan?
  var error: String

  var success: Bool {
    plan != nil && error.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
  }

  init(plan: AgentPlan? = nil, error: String = "") {
    self.plan = plan
    self.error = error
  }
}

enum AgentPlanEditor {
  static func inputKey(action: AgentAction) -> String? {
    switch action.kind {
    case .callConnector, .controlDevice:
      return "prompt"
    case .typeText, .createNotification:
      return "text"
    case .setAlarm:
      return "message"
    default:
      return nil
    }
  }

  static func inputValue(action: AgentAction) -> String {
    guard let key = inputKey(action: action) else {
      return ""
    }
    return action.parameters[key] ?? ""
  }

  static func updatePendingAction(
    plan: AgentPlan,
    actionId: String,
    description: String,
    input: String
  ) -> AgentPlanEditResult {
    guard let action = plan.actions.first(where: { $0.id == actionId }) else {
      return failure("Action is no longer in the active plan")
    }
    guard isEditablePending(action) else {
      return failure("Only pending actions can be edited")
    }
    let cleanDescription = description.trimmed().clamped(to: maxDescriptionCharacters)
    guard !cleanDescription.isEmpty else {
      return failure("Action description cannot be empty")
    }
    let key = inputKey(action: action)
    let cleanInput = input.trimmed().clamped(to: maxInputCharacters(action.kind))
    if key != nil && cleanInput.isEmpty {
      return failure("Action input cannot be empty")
    }
    var updatedAction = action
    updatedAction.description = cleanDescription
    if let key {
      updatedAction.parameters[key] = cleanInput
    }
    return validateEditedPlan(
      original: plan,
      actions: plan.actions.map { $0.id == actionId ? updatedAction : $0 },
      editSummary: "updated:\(actionId)"
    )
  }

  static func removePendingAction(plan: AgentPlan, actionId: String) -> AgentPlanEditResult {
    guard let action = plan.actions.first(where: { $0.id == actionId }) else {
      return failure("Action is no longer in the active plan")
    }
    guard isEditablePending(action) else {
      return failure("Only pending actions can be removed")
    }
    guard plan.actions.count > 1 else {
      return failure("A plan must contain at least one action")
    }
    if let dependent = plan.actions.first(where: { dependencyIds($0).contains(actionId) }) {
      return failure("Remove dependent action \(dependent.description) first")
    }
    return validateEditedPlan(
      original: plan,
      actions: plan.actions.filter { $0.id != actionId },
      editSummary: "removed:\(actionId)"
    )
  }

  static func movePendingAction(plan: AgentPlan, actionId: String, offset: Int) -> AgentPlanEditResult {
    guard [-1, 1].contains(offset) else {
      return failure("Unsupported move")
    }
    guard let currentIndex = plan.actions.firstIndex(where: { $0.id == actionId }) else {
      return failure("Action is no longer in the active plan")
    }
    let action = plan.actions[currentIndex]
    guard isEditablePending(action) else {
      return failure("Only pending actions can be moved")
    }
    let targetIndex = currentIndex + offset
    guard plan.actions.indices.contains(targetIndex) else {
      return failure("Action is already at the plan boundary")
    }
    guard isEditablePending(plan.actions[targetIndex]) else {
      return failure("Completed or running actions cannot be reordered")
    }
    var reordered = plan.actions
    let moved = reordered.remove(at: currentIndex)
    reordered.insert(moved, at: targetIndex)
    return validateEditedPlan(
      original: plan,
      actions: reordered,
      editSummary: "moved:\(actionId):\(offset)"
    )
  }

  private static func validateEditedPlan(
    original: AgentPlan,
    actions: [AgentAction],
    editSummary: String
  ) -> AgentPlanEditResult {
    var candidate = original
    candidate.actions = actions
    candidate.revision = original.revision + 1
    candidate.routeRationale = original.routeRationale.components(separatedBy: " User edit:").first.orEmpty +
      " User edit: \(editSummary)."
    candidate.validation = AgentPlanValidator.validate(candidate)
    if candidate.validation.valid {
      return AgentPlanEditResult(plan: candidate)
    }
    return failure(candidate.validation.issues.joined(separator: ", ").clamped(to: 300))
  }

  private static func isEditablePending(_ action: AgentAction) -> Bool {
    [.proposed, .pendingConfirmation].contains(action.status)
  }

  private static func dependencyIds(_ action: AgentAction) -> Set<String> {
    Set(listParameter("depends_on", action: action))
  }

  private static func listParameter(_ key: String, action: AgentAction) -> [String] {
    (action.parameters[key] ?? "")
      .split(separator: ",")
      .map { String($0).trimmed() }
      .filter { !$0.isEmpty }
  }

  private static func maxInputCharacters(_ kind: AgentActionKind) -> Int {
    switch kind {
    case .typeText:
      return 2_000
    case .createNotification:
      return 1_000
    case .setAlarm:
      return 200
    default:
      return 4_000
    }
  }

  private static func failure(_ message: String) -> AgentPlanEditResult {
    AgentPlanEditResult(error: message)
  }

  private static let maxDescriptionCharacters = 300
}

private extension Optional where Wrapped == String {
  var orEmpty: String {
    self ?? ""
  }
}

private extension String {
  func trimmed() -> String {
    trimmingCharacters(in: .whitespacesAndNewlines)
  }

  func clamped(to limit: Int) -> String {
    String(prefix(max(limit, 0)))
  }
}

File: apps/android/app/src/main/java/com/signalasi/chat/AgentPlanEditor.kt
package com.signalasi.chat

data class AgentPlanEditResult(
    val plan: AgentPlan? = null,
    val error: String = ""
) {
    val success: Boolean get() = plan != null && error.isBlank()
}

object AgentPlanEditor {
    fun inputKey(action: AgentAction): String? = when (action.kind) {
        AgentActionKind.CALL_CONNECTOR,
        AgentActionKind.CONTROL_DEVICE -> "prompt"
        AgentActionKind.TYPE_TEXT -> "text"
        AgentActionKind.CREATE_NOTIFICATION -> "text"
        AgentActionKind.SET_ALARM -> "message"
        else -> null
    }

    fun inputValue(action: AgentAction): String = inputKey(action)
        ?.let { action.parameters[it] }
        .orEmpty()

    fun updatePendingAction(
        plan: AgentPlan,
        actionId: String,
        description: String,
        input: String
    ): AgentPlanEditResult {
        val action = plan.actions.firstOrNull { it.id == actionId }
            ?: return failure("Action is no longer in the active plan")
        if (!action.isEditablePending()) return failure("Only pending actions can be edited")
        val cleanDescription = description.trim().take(MAX_DESCRIPTION_CHARACTERS)
        if (cleanDescription.isBlank()) return failure("Action description cannot be empty")
        val key = inputKey(action)
        val cleanInput = input.trim().take(maxInputCharacters(action.kind))
        if (key != null && cleanInput.isBlank()) return failure("Action input cannot be empty")
        val updatedAction = action.copy(
            description = cleanDescription,
            parameters = if (key == null) action.parameters else action.parameters + (key to cleanInput)
        )
        return validateEditedPlan(
            plan,
            plan.actions.map { if (it.id == actionId) updatedAction else it },
            "updated:$actionId"
        )
    }

    fun removePendingAction(plan: AgentPlan, actionId: String): AgentPlanEditResult {
        val action = plan.actions.firstOrNull { it.id == actionId }
            ?: return failure("Action is no longer in the active plan")
        if (!action.isEditablePending()) return failure("Only pending actions can be removed")
        if (plan.actions.size <= 1) return failure("A plan must contain at least one action")
        val dependent = plan.actions.firstOrNull { actionId in it.dependencyIds() }
        if (dependent != null) return failure("Remove dependent action ${dependent.description} first")
        return validateEditedPlan(
            plan,
            plan.actions.filterNot { it.id == actionId },
            "removed:$actionId"
        )
    }

    fun movePendingAction(plan: AgentPlan, actionId: String, offset: Int): AgentPlanEditResult {
        if (offset !in setOf(-1, 1)) return failure("Unsupported move")
        val currentIndex = plan.actions.indexOfFirst { it.id == actionId }
        if (currentIndex < 0) return failure("Action is no longer in the active plan")
        val action = plan.actions[currentIndex]
        if (!action.isEditablePending()) return failure("Only pending actions can be moved")
        val targetIndex = currentIndex + offset
        if (targetIndex !in plan.actions.indices) return failure("Action is already at the plan boundary")
        if (!plan.actions[targetIndex].isEditablePending()) {
            return failure("Completed or running actions cannot be reordered")
        }
        val reordered = plan.actions.toMutableList().apply {
            val moved = removeAt(currentIndex)
            add(targetIndex, moved)
        }
        return validateEditedPlan(plan, reordered, "moved:$actionId:$offset")
    }

    private fun validateEditedPlan(
        original: AgentPlan,
        actions: List<AgentAction>,
        editSummary: String
    ): AgentPlanEditResult {
        val revision = original.revision + 1
        val retainedActionIds = actions.mapTo(hashSetOf(), AgentAction::id)
        val removedActions = original.actions
            .filterNot { action -> action.id in retainedActionIds }
            .map { action -> action.supersededByPlanRevision(original.revision, revision) }
        val durableHistory = AgentProjectHistoryRetentionPolicy.latestSnapshots(
            original.actionHistory + removedActions
        ).takeLast(AgentLongTaskPersistenceLimits.MAX_ACTIONS)
        val revisedActions = actions.map { action ->
            if (action.status in EDITABLE_ACTION_STATUSES) {
                action.withPlanRevision(revision)
            } else {
                action.ensurePlanRevision(original.revision)
            }
        }
        var candidate = original.copy(
            actions = revisedActions,
            actionHistory = durableHistory,
            revision = revision,
            routeRationale = original.routeRationale.substringBefore(" User edit:") +
                " User edit: $editSummary."
        )
        candidate = candidate.copy(validation = AgentPlanValidator.validate(candidate))
        return if (candidate.validation.valid) {
            AgentPlanEditResult(plan = candidate)
        } else {
            failure(candidate.validation.issues.joinToString(", ").take(300))
        }
    }

    private fun AgentAction.isEditablePending(): Boolean = status in EDITABLE_ACTION_STATUSES

    private val EDITABLE_ACTION_STATUSES = setOf(
        AgentActionStatus.PROPOSED,
        AgentActionStatus.PENDING_CONFIRMATION
    )

    private fun maxInputCharacters(kind: AgentActionKind): Int = when (kind) {
        AgentActionKind.TYPE_TEXT -> 2_000
        AgentActionKind.CREATE_NOTIFICATION -> 1_000
        AgentActionKind.SET_ALARM -> 200
        else -> 4_000
    }

    private fun failure(message: String): AgentPlanEditResult = AgentPlanEditResult(error = message)

    private const val MAX_DESCRIPTION_CHARACTERS = 300
}
